sdk/trace/batch_span_processor.go

```diff
@@ -167,32 +154,58 @@ func WithBlocking() BatchSpanProcessorOption {
 	}
 }
 
-// processQueue removes spans from the `queue` channel until there is
-// no more data.  It calls the exporter in batches of up to
-// MaxExportBatchSize until all the available data have been processed.
-func (bsp *BatchSpanProcessor) processQueue(batch *[]*export.SpanData) {
+// processQueue removes spans from the `queue` channel until processor
+// is shut down. It calls the exporter in batches of up to MaxExportBatchSize
+// waiting up to ScheduledDelayMillis to form a batch.
+func (bsp *BatchSpanProcessor) processQueue() {
+	ticker := time.NewTicker(bsp.o.ScheduledDelayMillis)
+	defer ticker.Stop()
+
+	batch := make([]*export.SpanData, 0, bsp.o.MaxExportBatchSize)
+
+	exportSpans := func() {
+		if len(batch) > 0 {
+			bsp.e.ExportSpans(context.Background(), batch)
+			batch = batch[:0]
+		}
+	}
+
+loop:
 	for {
-		// Read spans until either the buffer fills or the
-		// queue is empty.
-		for ok := true; ok && len(*batch) < bsp.o.MaxExportBatchSize; {
-			select {
-			case sd := <-bsp.queue:
-				if sd != nil && sd.SpanContext.IsSampled() {
-					*batch = append(*batch, sd)
+		select {
+		case <-bsp.stopCh:
+			break loop
+		case <-ticker.C:
+			exportSpans()
+		case sd := <-bsp.queue:
+			if sd.SpanContext.IsSampled() {
+				batch = append(batch, sd)
+				if len(batch) == bsp.o.MaxExportBatchSize {
+					ticker.Reset(bsp.o.ScheduledDelayMillis)
+					exportSpans()
 				}
-			default:
-				ok = false
 			}
 		}
+	}
 
-		if len(*batch) == 0 {
-			return
+	// Consume queue before close to unblock enqueue and prevent
+	// "panic: send on closed channel".
```

Comment:
This is not my idea - I believe it is how it worked previously - there is nothing else that prevents "panic: send on closed channel".

Reply:
I believe the former logic is trying to ensure that any span that finished before the processor is stopped will be flushed to the exporter.  If the call to `span.End()` has returned before stopping the processor, we should be sure to export that span. 

There is still a race condition--if the `queue` channel is closed a reader could panic. A concurrent call to `span.End()` might have loaded the processors but not enqueued yet. I would prefer to see a safe solution that leaks resources: perhaps a goroutine that reads indefinitely from the queue and drops the results.